File: dwcj-engine/src/main/java/org/dwcj/component/html/elements/Paragraph.java
package org.dwcj.component.html.elements;

import org.dwcj.component.Component;
import org.dwcj.component.element.annotation.NodeName;
import org.dwcj.component.html.HtmlComponentContainer;

/**
 * Component representing a {@code p} element.
 *
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/p">HTML p Tag</a>
 *
 * @author Hyyan Abo Fakher
 * @since 23.06
 */
// We're purposefully ignoring the deep inheritance warning here because we've designed our class
// hierarchy to meet the unique requirements of our UI framework. This design closely aligns with
// our framework's specific goals and emphasizes the need for caution when considering any changes.
//
// Any changes to the inheritance structure should be thoughtfully evaluated in the context of our
// framework's needs. The current structure is essential for meeting those needs.
@SuppressWarnings("squid:S110")
@NodeName("p")
public class Paragraph extends HtmlComponentContainer<Paragraph> {

  /**
   * Creates a new empty Paragraph.
   */
  public Paragraph() {
    super();
  }

  /**
   * Creates a new Paragraph with the given text.
   *
   * @param text the text
   */
  public Paragraph(String text) {
    super();
    setText(text);
  }

  /**
   * Creates a new Paragraph with the given child components.
   *
   * @param components the child components
   */
  public Paragraph(Component... components) {
    super(components);
  }
}

File: webforj-kotlin/src/main/kotlin/com/webforj/kotlin/dsl/component/html/elements/Paragraph.kt
package com.webforj.kotlin.dsl.component.html.elements

import com.webforj.component.html.elements.Paragraph
import com.webforj.concern.HasComponents
import com.webforj.kotlin.dsl.WebforjDsl
import com.webforj.kotlin.dsl.init

/**
 * Creates an `Paragraph` with an optional [text] value.
 * ```
 * ... {
 *  Paragraph() // Empty Paragraph element
 *  Paragraph("text") // Paragraph element with text
 * }
 * ```
 * @param text Optional text to add to the `Paragraph`.
 * @param block The initialization steps of the `Paragraph`.
 * @return The configured `Paragraph` instance.
 * @see Paragraph
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/p">HTML p Tag</a>
 */
fun @WebforjDsl HasComponents.paragraph(text: String? = null, block: @WebforjDsl Paragraph.() -> Unit = {}): Paragraph {
    val paragraph = text?.let { Paragraph(text) } ?: Paragraph()
    return init(paragraph, block)
}
